dotnet/Razorvine.Pyrolite/Pyrolite/Pickle/UnpicklerOfT.cs

```diff
@@ -0,0 +1,818 @@
+﻿using Razorvine.Pickle.Objects;
+using System;
+using System.Buffers.Binary;
+using System.Buffers.Text;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection;
+
+namespace Razorvine.Pickle
+{
+    // the following type is generic in order to allow for 
+    // IInputReader interface method devirtualizaiton and inlining
+    // please see https://adamsitnik.com/Value-Types-vs-Reference-Types/#how-to-avoid-boxing-with-value-types-that-implement-interfaces for more
+    internal class Unpickler<T> where T : struct, IInputReader
```

Comment:
Why a class rather than a struct?  Just for simplicity and the extra allocation doesn't matter?

Reply:
it has too many fields (however it's not send anywhere) + simplicity